### PR TITLE
Earth - Schanen

### PR DESCRIPTION
**Pull request by schanenR**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | Session information is stored in users browser via a cache. It is clears when the browser is closed. Flash is used to generate messages in response to user requests and they are cleared when a new request is made.
What was one thing that you gained more clarity on through this assignment? | Relationships, specifically indirect relationships. 
What is the Heroku URL of your deployed application? | 

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+Rails.application.routes.draw do
+
+  root to: 'works#top_works'
+
+  resources :votes, only: [:index, :new, :create]
```
Comment (CheezItMan) [inline config/routes.rb]: I suggest nesting this under works, and you only need to create votes, not list them or have a new form.

```ruby
resources :works do
  resources :votes, only: [:create]
end
```

File: config/routes.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+  resources :works
+  get '/works/top_works', to: 'works#top_works', as: 'top_works'
```
Comment (CheezItMan) [inline config/routes.rb:10]: Why not have an index methods for works and leave it at that?

File: config/routes.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+  get '/login', to: 'users#login_form', as: 'login'
+  post '/login', to: 'users#login'
+  post '/logout', to: 'users#logout', as: 'logout'
+  get '/users/current', to: 'users#current', as: 'current_user'
+
+  resources :users, only: [:index, :show] do
+    resources :votes, only: [:index, :new]
+  end
```
Comment (CheezItMan) [inline config/routes.rb]: See above on votes.

File: app/views/works/_list_works.html.erb [hunk 4]
```diff
@@ -0,0 +1,29 @@
+
+
+<section class="media-table">
```
Comment (CheezItMan) [inline app/views/works/_list_works.html.erb]: This is a great idea, but does it have to be books?  Maybe you can pass in the parameters `category` and `works`, and You can print the category in the h4 here and use `works` in the loop.

File: app/views/works/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      <tr>
+        <td><%= work.votes.count %></td>
+        <td><%= link_to work.title, work_path(work.id) %></td>
+        <td><%= work.creator %></td>
+        <td><%= work.publication_date %></td>
+        <td class="button button-primary"><%= button_to 'Upvote', votes_path %></td>
+      </tr>
+      <% end %>
+    </tbody>
+  </table>
```
Comment (CheezItMan) [inline app/views/works/index.html.erb]: This would be a good place to use your partial, same for the other loops below.

File: app/views/works/top_works.html.erb [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        <li>
+
+          <h4>
+            <%= link_to movie.title, work_path(movie.id) %>
+            <small class="top-ten-creator">by <%= movie.creator %></small>
+          </h4>
+          <p><%= link_to movie.votes.count %> votes</p>
+        </li>
+      <% end %>
+    </ul>
```
Comment (CheezItMan) [inline app/views/works/top_works.html.erb:31]: This would also be a good place for the partial you had.

File: app/controllers/works_controller.rb [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    end
+  end
+
+  def destroy
+
+    if @work
+      @work.destroy
+      redirect_to works_path
+    else
+      head :not_found
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:76]: I suggest redirecting with a flash notice

File: app/controllers/works_controller.rb [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+  def update
+    if @work.nil?
+      redirect_to works_path
+      return
+    elsif @work.update(work_params)
+      redirect_to work_path(@work.id)
+      return
+    else
+      render :edit
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:65]:
```suggestion
      # Add flash notices here!!!
      render :edit, status: :bad_request
```

File: app/controllers/works_controller.rb [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+      flash.now[:error] = "Something happened. Work not added."
+      render :new, status: :bad_request
+      return
+    end
+  end
+
+  def edit
+
+    if @work.nil?
+      head :not_found
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:52]: I suggest redirecting and doing a flash notice

```suggestion
      flash[:error] = "Work #{params[:id]} is not found."
      redirect_to root_path
```

File: test/models/work_test.rb [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  #
+  #     # Assert
+  #     expect(new_driver.trips.count).must_equal 2
+  #     new_driver.trips.each do |trip|
+  #       expect(trip).must_be_instance_of Trip
+  #     end
+  #   end
+  # end
+
+  describe "validations" do
```
Comment (CheezItMan) [inline test/models/work_test.rb:24]: This is great!

**Review by CheezItMan — COMMENTED** [review]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ⚠️ Instead it renders head... 
| 4. Can edit this work successfully, and get a flash message | ⚠️ No flash notice
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️?
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔, It prevents the model from being created, but the validation errors are not shown...
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️, although a flash notice would be nice
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ⚠️ I can't vote!
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ⚠️ I can't vote!
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ⚠️ I can't vote!
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ⚠️ Not deployed to Heroku
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ⚠️ Still working on styling
| 3. Practices git with at least 25 small commits and meaningful commit messages | ⚠️, not many commits

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️, you have an extra route for the new action for votes (you don't need a form page for voting)
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️, no, but you're clearly trying.
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ⚠️, missing business logic methods
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️, yes with some work yet to do.  I also strongly suggest redirecting instead of using `head`.  Also include status codes when you render. 

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ⚠️ Only fixtures for Works
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ⚠️ No tests
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ⚠️, No tests
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ⚠️, Great tests for validations and no tests for relationships.
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ⚠️, no tests for business logic methods

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | ❤️





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Descriptive/Readable | ✅
| Logical/Organized | ✅

## Summary

This looks like it was a tough project for you Schanen.  You didn't have the voting mechanics working, but you did have the login functionality and the basic CRUD for works.  You also had controller filters working and were on the path toward using a good partial view.  There was a lot not working, but some good things here too.

A prioritized area for improvement:

1.  Writing good tests for model methods.
1.  Writing business logic methods using `.order` and `.where` to sort and filter collections of model instances.
1.  Looking for areas where it makes sense to use nested routes.  Like creating a vote for a work in this project.
